### PR TITLE
Filter by region and zone too when getting default instance type

### LINKED ISSUE #6290: [GCP] Cannot launch VM in GCP us-west2-a
Repro:

```bash
> sky launch --infra gcp/us-west2/us-west2-a
No resource satisfying GCP(, region=us-west2, zone=us-west2-a) on GCP.
sky.exceptions.ResourcesUnavailableError: Catalog does not contain any instances satisfying the request: 1x GCP(, region=us-west2, zone=us-west2-a).
To fix: relax or change the resource requirements.

Hint: sky show-gpus to list available accelerators.
      sky check to check the enabled clouds.
Hint: Check Per Resource Hint
```

But able to launch in `us-west2-c`:

```
Considered resources (1 node):
---------------------------------------------------------------------------------
 INFRA              INSTANCE        vCPUs   Mem(GB)   GPUS   COST ($)   CHOSEN
---------------------------------------------------------------------------------
 GCP (us-west2-c)   n4-standard-8   8       32        -      0.46          ✔
---------------------------------------------------------------------------------
```

Version:

```
sky api info
Using SkyPilot API server and dashboard: http://****
├── Status: healthy, commit: 5ed1775c386bbd0cc4b9d1c80fc0d2d91b352870, version: 1.0.0-dev0
└── User: aylei (b673d4fd)
```

### PR DESCRIPTION
**Pull request by kevinmingtarja**
Fixes #6290.

We recently added support for n4 instances for GCP ([PR](https://github.com/skypilot-org/skypilot/pull/6253)) and added it to `_DEFAULT_INSTANCE_FAMILY`, so now the default instance types for GCP are n2 and n4. And n4 is cheaper than n2 for most if not all zones.

The problem is, in `common.get_instance_type_for_cpus_mem_impl`, we only filter by CPU and memory, before sorting by price and getting the cheapest one.

We don't filter the zone, meaning we don't take into account the possibility that an instance may be cheapest overall, but doesn't exist in a particular zone, like what's happening here with n4-standard.

This PR fixes it by adding `region` and `zone` as optional input params to:
1. `cloud.Cloud.get_default_instance_type`
2. `catalog.get_default_instance_type`
3. `common.get_instance_type_for_cpus_mem_impl`, where we finally do `df = _filter_region_zone(df, region, zone)`

Tested (run the relevant ones):

- [x] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local)
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

### CODE REVIEW

File: sky/clouds/gcp.py [hunk 1]
```diff
@@ -681,7 +684,10 @@ def _get_feasible_launchable_resources(
             host_vm_type = GCP.get_default_instance_type(
                 cpus=resources.cpus,
                 memory=resources.memory,
-                disk_tier=resources.disk_tier)
+                disk_tier=resources.disk_tier,
+                region=resources.region,
+                zone=resources.zone)
+            logger.info(f'host_vm_type: {host_vm_type}')
```
Comment (aylei) [inline sky/clouds/gcp.py]: Shall we remove this log or make this more meaningful?

File: sky/catalog/common.py [hunk 2]
```diff
@@ -476,8 +476,11 @@ def _filter_region_zone(df: 'pd.DataFrame', region: Optional[str],
 
 
 def get_instance_type_for_cpus_mem_impl(
```
Comment (aylei) [inline sky/catalog/common.py:478]: Can we add a unit test for this function with mocked data? i.e. avoid depending on specific cloud catalog to repro this edge case

**Review by aylei — APPROVED** [review]
LGTM, thanks @kevinmingtarja !

File: sky/clouds/gcp.py (hunk 1 above)
Comment (kevinmingtarja) [reply sky/clouds/gcp.py]: Ah right, I forgot to remove this. Thanks for catching this!

File: sky/catalog/common.py (hunk 2 above)
Comment (kevinmingtarja) [reply sky/catalog/common.py:478]: Sounds good, I've added unit tests.

**Review by aylei — APPROVED** [review]
